styx-service-common/src/test/java/com/spotify/styx/storage/DatastoreStorageTransactionTest.java

```diff
@@ -104,175 +107,90 @@ public void tearDown() throws Exception {
   }
 
   @Test
-  public void shouldCommitEmptyTransaction() throws IOException {
-    final CheckedDatastoreTransaction transaction = datastore.newTransaction();
-    DatastoreStorageTransaction storageTransaction = new DatastoreStorageTransaction(transaction);
-    storageTransaction.commit();
-    assertFalse(transaction.isActive());
-  }
-
-  @Test
-  public void shouldThrowIfUnexpectedDatastoreError() throws IOException {
-    final CheckedDatastoreTransaction transaction = datastore.newTransaction();
-    DatastoreStorageTransaction storageTransaction = new DatastoreStorageTransaction(transaction);
-
-    transaction.rollback();
-    try {
-      storageTransaction.commit();
-      fail("Expected exception!");
-    } catch (TransactionException e) {
-      assertFalse(e.isConflict());
-    }
-  }
+  public void shouldRetryOnConflict() throws Exception {
+    var executor = Executors.newCachedThreadPool();
+
+    var workflow = TestData.WORKFLOW_WITH_RESOURCES;
+
+    // Store workflow
+    storage.runInTransactionWithRetries(tx -> {
+      tx.store(workflow);
+      return null;
+    });
+
+    // Start a losing transaction that reads, waits for barrier and then stores the workflow
```

Comment:
Comment more how they actually compete and the result is the first one will be executed twice.